### ISSUE TITLE
BigDecimal support compareTo method

### ISSUE THREAD

**Issue by sergioszy** [beanshell/beanshell#569]
Hi,
only for compatibility is it possible to add the compareTo method to bsh.Primitive

```java
    /**
        CompareTo
    */
    public int compareTo( Object obj )
    {
        if ( !( obj instanceof Primitive ) )
            if ( wrapperMap.containsKey(obj.getClass()) )
                obj = new Primitive(obj);
            else
                return -1;
        Primitive pobj = (Primitive) obj;
        if ( pobj.isNumber() && this.isNumber() ) {
            if ( this.value instanceof BigDecimal )
                return ((BigDecimal)this.value).compareTo((BigDecimal)castNumber(BigDecimal.class, pobj.numberValue()));
            if ( pobj.value instanceof BigDecimal )
                return ((BigDecimal)pobj.value).compareTo((BigDecimal)castNumber(BigDecimal.class, this.numberValue())) * ( -1);
            if (Types.isFloatingpoint(this.value) || Types.isFloatingpoint(pobj.value))

                return ( this.numberValue().doubleValue() == pobj.numberValue().doubleValue() )? 0 : ( this.numberValue().doubleValue() > pobj.numberValue().doubleValue() )? 1 : -1;

            if ( this.value instanceof BigInteger )
                return ((BigInteger)this.value).compareTo((BigInteger)castNumber(BigInteger.class, pobj.numberValue()));
            if ( pobj.value instanceof BigInteger )
                return ((BigInteger)pobj.value).compareTo((BigInteger)castNumber(BigInteger.class, this.numberValue())) * ( -1);

            return ( this.numberValue().longValue() == pobj.numberValue().longValue() )? 0 :
                     ( this.numberValue().longValue() > pobj.numberValue().longValue() )? 1 : -1;

        }
        return String.valueOf(this.value).compareTo( String.valueOf( pobj.value ));
    }

```

**Comment by sergioszy** [comment]
After a little thought, I added,
add, subtract, multiply and divide
attached as is the Primitive class

[Primitive.java.txt](https://github.com/beanshell/beanshell/files/3741667/Primitive.java.txt)

**Comment by nickl-** [comment]
@sergioszy I am not quite understanding the purpose for your change suggestion.

Firstly what I assume you are suggesting is to have the `Primitive` class implement `Comparable`, this is the purpose for a `compareTo` method right?

All wrapper types are already comparable just like `BigDecimal`, which should greatly simplify your implementation should we want to make `Primitive` comparable.

The thing is that the `Primitive` class is for internal use, it is a way for us to retain the unwrapped functionality of java primitives. Just as with java, in BeanShell, we auto box primitives wherever objects are used, for example adding an `int` to a collection will become an `Integer` and sorting then works as a result of its comparability.


```java
bsh % List l = {6, 2, 5, 3};
--> $5 = [6I, 2I, 5I, 3I] :List
bsh % Collections.sort(l);
--> void
bsh % l;
--> $6 = [2I, 3I, 5I, 6I] :List
```

It is not intended to treat primitives as objects however.

Is there another reason I am not seeing why `Primitive` should be comparable?

----------

The reason we have included the `BigDecimal` and `BigInteger` is to be able to treat large numbers as primitives, which java cannot.

```java
bsh % 3768767687687687687687687687687686 + 3;
--> $1 = 3768767687687687687687687687687689W :BigInteger
```

```java
jshell> 3768767687687687687687687687687686 + 3;
|  Error:
|  integer number too large: 3768767687687687687687687687687686
|  3768767687687687687687687687687686 + 3;
|  ^
```

We treat these classes as wrapper types meaning all math operators works as per normal for wrapper types, which java cannot.

```java
bsh % new BigInteger("3768767687687687687687687687687686") + new Integer(3);
--> $3 = 3768767687687687687687687687687689W :BigInteger
```

```java
jshell> new BigInteger("3768767687687687687687687687687686") + new Integer(3);
|  Error:
|  bad operand types for binary operator '+'
|    first type:  java.math.BigInteger
|    second type: java.lang.Integer
|  new BigInteger("3768767687687687687687687687687686") + new Integer(3);
|  ^-------------------------------------------------------------------^
```

Can you please motivate why we should include math functions to the `Primitive` class?

**Comment by sergioszy** [comment]
Hi Nick,
I suggested it for compatibility with the original beanshell and for java compatibility.
With version 3 it is not possible to execute the following logic:
if (var1.compareTo (var2) == 0 ) {
     .....
}
it is also not possible to execute this
varX = var1.add (var2) .multiply (var3) .divide (var4);

for this reason also add the methods 
[methods.txt](https://github.com/beanshell/beanshell/files/3840820/methods.txt)

add / subtract / multiply / divide

But if possible with version 2.b04 and with java.

with this simple addition to the Primitive class, everything
this is possible and I don't see that it affects stability and
performance of the new BeanShell

**Comment by opeongo** [comment]
> 
> 
> Hi Nick,
> I suggested it for compatibility with the original beanshell and for java compatibility.
> With version 3 it is not possible to execute the following logic:
> if (var1.compareTo (var2) == 0 ) {
> .....
> }

Compatibility with Java seems to be a good idea. +1 on this.


> it is also not possible to execute this
> varX = var1.add (var2) .multiply (var3) .divide (var4);

Is there an existing class in the JDK that supports these methods, or is this a BeanShell only convenience.  If it is something that is already in the JDK that we are trying to be compatible with then +1, but if it is a new convenience then I don't see that it has much value.

**Comment by nickl-** [comment]
This may be a duplicate of #609 

We need to check if these arithmetic functions are from BigDecimal then 609 will resolve this too

**Comment by nickl-** [comment]
Closed in favour of #609 and once `Primitive` exposes the value instance `compareTo` from `Number` and all the `java.math` functionality will be available.
